core/src/main/scala/kafka/coordinator/transaction/TransactionCoordinator.scala

```diff
@@ -552,6 +560,22 @@ class TransactionCoordinator(brokerId: Int,
     }
   }
 
+  // Fence the producer by either incrementing the producer epoch or, if the epoch is exhausted, rotating the producer ID.
+  // Rotating the producer ID is necessary because we can't always write the metadata transition to the transaction log
+  // (such as in the case when there are too few in-sync replicas), but we always reflect the fencing in the transaction
+  // metadata. If we bump the epoch to the point of exhaustion but fail to write to the log, we would be stuck
+  // with an Ongoing transaction that couldn't be closed by the coordinator because our epoch is exhausted and couldn't
+  // be closed by the client because it would be stuck without a valid producer ID and epoch.
+  private def prepareFenceProducer(txnMetadata: TransactionMetadata): TxnTransitMetadata = {
+    if (txnMetadata.isProducerEpochExhausted) {
+      val newProducerId = producerIdManager.generateProducerId()
+      txnMetadata.prepareProducerIdRotation(newProducerId, txnMetadata.txnTimeoutMs, time.milliseconds(),
```

Comment:
I'm trying to follow the logic here. Suppose we have an ongoing transaction that will be aborted after this call returns. We need to transition to PrepareAbort and we need to send markers using the old producerId and `Short.MaxValue` as the epoch. Does this logic ensure that that happens?

Reply:
You're right, it doesn't. This would actually write the abort marker with the new producer ID, which we don't want. After thinking more about that requirement, I don't think there's a good way to do this by rotating the producer ID. We would need to be able to keep the current producer ID around in the metadata until the underlying unavailability cleared, which could even span multiple producer ID rotations. I'm going to push a new approach to this fix that instead rolls the epoch bump back when the write fails.

Reply:
Thanks for the updates @bob-barrett . I have another question about the new approach: does that mean the write may be appended locally with the new epoch, and then afterwards we roll back to a smaller epoch. If that's the case, is it possible we may fall into KAFKA-9144 scenario if the isr shrinks and we do not have min.isr > 1?

Reply:
@guozhangwang If the write succeeds locally, we will never roll the epoch back. This rollback logic only triggers if the local write fails. Once that write succeeds, the epoch has been bumped it would be unsafe to roll back. And if the write fails, we will never send the marker to the partition, so there is no chance to ever persisting the higher epoch.